exchange/utils_test.go

```diff
@@ -2543,3 +2543,23 @@ func TestBuildXPrebidHeader(t *testing.T) {
 		assert.Equal(t, test.result, result, test.description+":result")
 	}
 }
+
+func TestSourceExtSChainCopied(t *testing.T) {
+	bidRequest := &openrtb2.BidRequest{}
+	bidderSchains := map[string]*openrtb_ext.ExtRequestPrebidSChainSChain{
+		"bidder1": {
+			Complete: 1,
+		},
+		"bidder2": {
+			Complete: 2,
+		},
+	}
+
+	copy1 := *bidRequest
+	prepareSource(&copy1, "bidder1", bidderSchains)
+	copy2 := *bidRequest
+	prepareSource(&copy2, "bidder2", bidderSchains)
```

Comment:
The second call to `prepareSource(...)` seems to test the exact same scenario than the first. Should we remove?

Reply:
No, that would destroy the purpose of the test. It’s supposed to test that a sequence of calls to prepareSource will only modify the bidRequest being sent in. Maybe I wasn’t clear about what this bug is all about. 

Without the big fix “copy1” would be affected by the second call with the “copy2” parameter. 

This causes bidder specific schains to not work at all and has probably never worked.

Reply:
i.e. the following would pass(which is clearly wrong):

assert.Equal(t, json.RawMessage(`{"schain":{"complete":2,"nodes":null,"ver":""}}`), copy1.Source.Ext, "First schain was overwritten or not set")
	assert.Equal(t, json.RawMessage(`{"schain":{"complete":2,"nodes":null,"ver":""}}`), copy2.Source.Ext, "Second schain was overwritten or not set")

Reply:
If newly added lines 307 to 309 get removed `TestSourceExtSChainCopied` still succeeds, and I believe this happens because in both calls to `prepareSource()` we pass a `bidRequest` with a `nil` `Source` field. In other words, both calls to `prepareSource()` seem to take the true branch of the `if` statement in line 305:
```
304     // set source
305     if req.Source == nil {
306         req.Source = &openrtb2.Source{}
307     } else {
308         sourceCopy := *req.Source
309         req.Source = &sourceCopy
310     }
exchange/utils.go
```
And the newly added `false` branch (lines 308 and 309) is not covered by your `TestSourceExtSChainCopied`. Can you modify?

> Without the bug fix “copy1” would be affected by the second call with the “copy2” parameter.

Can we explicitly assert this?

Reply:
@guscarreon You're right, the test was sloppy. Let me fix this.

Reply:
@guscarreon please check latest commit